inc/az_ulib_registry_api.h

```diff
@@ -0,0 +1,198 @@
+ *                                                the system level.
+ *      @retval #AZ_ERROR_ULIB_BUSY               If the resources necessary for the
+ *                                                `az_ulib_registry_add` operation are busy.
+ *      @retval #AZ_ERROR_OUT_OF_MEMORY           If the flash space for `az_ulib_registry_add`
+ *                                                is not enough for a new registry entry.
+ */
+AZ_NODISCARD az_result az_ulib_registry_add(az_span key, az_span value);
+
+/**
+ * @brief   This function removes an #az_span key and its corresponding #az_span value from the
+ * device registry.
+ *
+ * This function goes through the registry and marks the node with the given key value pair as
+ * deleted so that it can no longer be accessible.
+ *
+ * @param[in]   key                 The #az_span key to remove from the registry.
+ *
+ * @pre         Registry shall already be initialized.
+ * @pre         \p key              shall not be `#AZ_SPAN_EMPTY`.
+ *
+ * @return The #az_result with the result of the registry operations.
+ *      @retval #AZ_OK                            If removing a value from the registry was
+ *                                                successful.
+ *      @retval #AZ_ERROR_ULIB_SYSTEM             If the `az_ulib_registry_delete` operation failed
+ *                                                on the system level.
+ *      @retval #AZ_ERROR_ULIB_BUSY               If the resources necessary for the
+ *                                                `az_ulib_registry_delete` operation are busy.
+ */
+AZ_NODISCARD az_result az_ulib_registry_delete(az_span key);
+
+/**
+ * @brief   This function initializes the device registry.
+ *
+ * This function initializes components that the registry needs upon reboot. This function is not
+ * thread safe and all other APIs shall only be invoked after the initialization ends.
+ *
+ * @param[in]   registry_cb         The pointer to #az_ulib_registry_control_block with the control
+ *                                  block that contains the registry memory.
+ *
+ * @pre         Registry shall **not** be initialized.
+ */
+void az_ulib_registry_init(const az_ulib_registry_control_block* registry_cb);
+
+/**
+ * @brief   This function deinitializes the device registry.
+ *
+ * This function deinitializes components that the registry used. The registry can be reinitialized.
+ * This function is not thread safe and all other APIs shall release the resource before calling
+ * the deinit() function.
+ *
+ * @pre         Registry shall already be initialized.
+ */
+void az_ulib_registry_deinit(void);
+
+/**
+ * @brief   Erase all memory reserved for registry.
+ *
+ * This function delete all configurations stored in the registry.
+ *
+ * @note    **There is no rollback for this operation.**
+ *
+ * @pre         Registry shall already be initialized.
+ */
+void az_ulib_registry_clean_all(void);
+
```

Comment:
consider adding a manual defrag API that allows users to defrag memory before the registry is full.

Reply:
Yes, as soon as we have a way to do that. :)

Reply:
#Wontfix